### PR TITLE
Fix S2094 FP: Allow to have empty class with non-empty ctor for base class

### LINKED ISSUE #6953: Fix S2094 FP: Records which set the base record's properties in their initializer
### Description

A C# record can specify a base record and set some of its properties in the call to the base constructor.

### Repro steps

```
public abstract record RuleApplyResult(bool Success);

public record RuleApplyDone() : RuleApplyResult(true);                  // FP - RuleApplyDone is a more specific version of RuleApplyResult

public record RuleApplySkipped(string Reason) : RuleApplyResult(false); // FP - RuleApplySkipped is a more specific version of RuleApplyResult
```

### Expected behavior

No issue should be raised for the descendant records.

### Actual behavior

An issue is raised for the descendant (more specific) record.

### Related information

v8.54

### PR DESCRIPTION
**Pull request by Corniel**
Fixes #6953

### CODE REVIEW

File: analyzers/src/SonarAnalyzer.CSharp/Rules/ClassShouldNotBeEmpty.cs [hunk 1]
```diff
@@ -51,7 +51,17 @@ public sealed class ClassShouldNotBeEmpty : ClassShouldNotBeEmptyBase<SyntaxKind
...
+    private static bool IsParameterlessRecord(SyntaxNode node) =>
+        RecordDeclarationSyntax(node) is { ParameterList.Parameters.Count: 0 } declaration
+        && !ProvidesBaseWithParameters(declaration);
+
+    private static bool ProvidesBaseWithParameters(RecordDeclarationSyntaxWrapper node) =>
+        node.BaseList?.Types.FirstOrDefault() is { } type
+        && PrimaryConstructorBaseTypeSyntaxWrapper.IsInstance(type)
+        && (PrimaryConstructorBaseTypeSyntaxWrapper)type is { ArgumentList.Arguments.Count: >= 1 };
+
+    private static RecordDeclarationSyntaxWrapper? RecordDeclarationSyntax(SyntaxNode node) =>
```
Comment (zsolt-kolbay-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/ClassShouldNotBeEmpty.cs:63]: This method isn't really needed anymore. All it does is check if the node is a record declaration or not. Moving that to `IsParameterlessRecord` will make the code less cluttered.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/ClassShouldNotBeEmpty.CSharp9.cs [hunk 2]
```diff
@@ -1,44 +1,61 @@
 ﻿using System;
 
-record EmptyRecord1();                                      // Noncompliant {{Remove this empty record, write its code or make it an "interface".}}
-//     ^^^^^^^^^^^^
-record EmptyRecord2() { };                                  // Noncompliant
+namespace Compliant
```
Comment (zsolt-kolbay-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/ClassShouldNotBeEmpty.CSharp9.cs:3]: It's fine to reorganize the test cases by grouping the code using namespaces, but
- if you do it in one of the test files then do it for the others as well for S2094, for the sake of consistency
- next time please do it in a separate commit. That will make it easier for us to review the PR, because we'd see what are the places where only the indentation has changed and what is completely new.

**Review by zsolt-kolbay-sonarsource — CHANGES_REQUESTED** [review]
Nicely done. Only some minor comments.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/ClassShouldNotBeEmpty.cs (hunk 1 above)
Comment (Corniel) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/ClassShouldNotBeEmpty.cs:58]: I've moved some stuff around. I think it improves the readability.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/ClassShouldNotBeEmpty.CSharp9.cs (hunk 2 above)
Comment (Corniel) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/ClassShouldNotBeEmpty.CSharp9.cs:3]: I did not want to put effort in rewriting all yet, as I wanted agreement first. ;)

Sorry for not putting it in a different commit.

Comment (Corniel) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/ClassShouldNotBeEmpty.CSharp9.cs:3]: @zsolt-kolbay-sonarsource I think I addressed all issues you raised?

**Review by zsolt-kolbay-sonarsource — APPROVED** [review]
LGTM
Kudos for reorganizing the test cases!
There are some merge conflicts that need to be resolved.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/ClassShouldNotBeEmpty.CSharp9.cs (hunk 2 above)
Comment (zsolt-kolbay-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/ClassShouldNotBeEmpty.CSharp9.cs:3]: Apologies. Didn't see the notification.

**Comment by Corniel** [comment]
@zsolt-kolbay-sonarsource Rebase done.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/ClassShouldNotBeEmpty.CSharp9.cs (hunk 2 above)
Comment (zsolt-kolbay-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/ClassShouldNotBeEmpty.CSharp9.cs:3]: /azp run [Sonar.Net](http://sonar.net/)

**Review by andrei-epure-sonarsource — CHANGES_REQUESTED** [review]
thanks @Corniel - I've left some last comments

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/ClassShouldNotBeEmpty.CSharp10.cs [hunk 3]
```diff
@@ -1,13 +1,19 @@
...
-record class NotEmptyRecordClass2()
+    record class EmptyRecordClass();                  // Noncompliant {{Remove this empty record, write its code or make it an "interface".}}
+    //           ^^^^^^^^^^^^^^^^
+    record class EmptyRecordClassWithEmptyBody() { }; // Noncompliant
+}
+namespace Ignore
 {
-    int RecordMember => 42;
-};
+    record struct EmptyRecordStruct();                // Compliant - this rule only deals with classes
```
[the remark this answers is not in the document]
Comment (Corniel) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/ClassShouldNotBeEmpty.CSharp10.cs:18]: Sorry, that was not intended.

**Review by andrei-epure-sonarsource — APPROVED** [review]
LGTM!

(note: I didn't check in detail each of the test cases that they got moved, I just sampled some)

**Comment by Corniel** [comment]
@zsolt-kolbay-sonarsource while reviewing some code with a collegae, I spotted a false negative for this rule:

``` C#
record EmptyChildWithoutBrackets : EmptyRecord; // Noncompliant
```

While at it, I extended the rule to also cover this.

**Comment by Corniel** [comment]
@zsolt-kolbay-sonarsource Can you do that for me please?
